javascript/Spread_operator.md

```diff
@@ -0,0 +1,171 @@
+﻿# Spread Operator in javascript
+
+## Projected time
+
+About 25-30 minutes
+
+- 10 min for destructuring
+- 20 min for examples
+
+## Spread operator
+Spread operator allows an iterable such as strings,array and sets to spread or expand individually inside a receiver.
+
+The **syntax** of spread operator is var variable  = [...value].
+
+## Why it is useful?
++ We can merge two objects together using spread operator.
+Example:
+```
+let obj1 = {a: 1, b: 2, c: 3}
+
+let obj2 = {d: 4, e: 5, f: 6}
+
+let merge = {...obj1, ...obj2} // merge is {a: 1, b: 2, c: 3, d: 4, e: 5, f: 6}
+```
++ Spread operator with object is used to make copy of existing object or to make a new object with properties.
+
+## Examples
+
+### Simple object
+
+Given this ```todoApp``` object we can use the object spread syntax and its properties into variables.
+
+```javaScript
+function todoApp(state = initialState, action) {
+
+  switch (action.type) {
+  
+    case SET_VISIBILITY_FILTER:
+    
+      return { ...state, visibilityFilter: action.filter }
+      
+    default:
+    
+      return state
+  }
+}
+```
+
+#### Code without using Spread operator
+
+```javaScript 
+function todoApp(state = initialState, action) {
+
+  switch (action.type) {
+  
+    case SET_VISIBILITY_FILTER:
+    
+      return Object.assign({}, state, {
+      
+        visibilityFilter: action.filter
+        
+      })
+     
+    default:
+    
+      return state
+  }
+}
+```
+### Arrays
+
+#### normal array concat() method 
+
+```javaScript
+let arr = [1,2,3]; 
+let arr2 = [4,5]; 
+  
+arr = arr.concat(arr2); 
+  
+console.log(arr); // [ 1, 2, 3, 4, 5 ] 
+```
+
+#### spread operator doing the concat job 
+
+```javaScript
+let arr = [1,2,3]; 
+let arr2 = [4,5]; 
+  
+arr = [...arr,...arr2]; 
+console.log(arr); // [ 1, 2, 3, 4, 5 ] 
+```
+
+#### spread operator for copying 
+
+```javaScript 
+let arr = ['a','b','c']; 
+let arr2 = [...arr]; 
+  
+console.log(arr); // [ 'a', 'b', 'c' ] 
+  
+arr2.push('d'); //inserting an element at the end of arr2 
+  
+console.log(arr2); // [ 'a', 'b', 'c', 'd' ] 
+console.log(arr); // [ 'a', 'b', 'c' ] 
+```
+
+#### copying without the spread operator 
+
+```javaScript
+let arr = ['a','b','c']; 
+let arr2 = arr; 
+  
+console.log(arr2); // [ 'a', 'b', 'c' ] 
+```
```

Comment:
The code from line 72-114 is copied as it is from https://www.geeksforgeeks.org/javascript-spread-operator/
You need to cite the sources.